# Data wrangling project

### Context
Wrangle [WeRateDogs](https://twitter.com/dog_rates) Twitter data to create interesting and trustworthy analyses and visualizations. The Twitter archive is great, but it only contains very basic tweet information. Additional gathering, then assessing and cleaning is required for "Wow!"-worthy analyses and visualizations.

In [7]:
import pandas as pd
import numpy as np
import requests
import tweepy
import json

import twitter_credetials as twc

## Gathering Data

In [8]:
twitter = pd.read_csv('twitter-archive-enhanced.csv')

In [9]:
url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'

response = requests.get(url)
with open('image_predictions.tsv', mode='wb') as file:
        file.write(response.content)

In [10]:
image_prd = pd.read_csv('image_predictions.tsv', sep='\t')

In [11]:
auth = tweepy.OAuthHandler(twc.consumer_key, twc.consumer_secret)
auth.set_access_token(twc.access_token, twc.access_token_secret)
api = tweepy.API(auth)

In [12]:
with open('tweet_json.txt', 'w') as file:
    for tweet_id in twitter['tweet_id']:
        try:
            json.dump(api.get_status(tweet_id, wait_on_rate_limit=True)._json, file)
            file.write('\n')
        except tweepy.TweepError as e:
            print('{} : {}'.format(e.args[0][0]['message'], tweet_id))

No status found with that ID. : 888202515573088257
No status found with that ID. : 873697596434513921
No status found with that ID. : 872668790621863937
No status found with that ID. : 872261713294495745
No status found with that ID. : 869988702071779329
No status found with that ID. : 866816280283807744
No status found with that ID. : 861769973181624320
No status found with that ID. : 856602993587888130
No status found with that ID. : 851953902622658560
No status found with that ID. : 845459076796616705
No status found with that ID. : 844704788403113984
No status found with that ID. : 842892208864923648
No status found with that ID. : 837366284874571778
No status found with that ID. : 837012587749474308
No status found with that ID. : 829374341691346946
No status found with that ID. : 827228250799742977
No status found with that ID. : 812747805718642688
No status found with that ID. : 802247111496568832
No status found with that ID. : 775096608509886464
No status found with that ID. :

In [13]:
df = []
with open('tweet_json.txt', 'r') as file:
    for line in file:
        tweet_dict = dict(json.loads(line))
        tweet_id = tweet_dict['id']
        favorite = tweet_dict['favorite_count']
        retweet = tweet_dict['retweet_count']
        
        df.append({'tweet_id': tweet_id,
                   'favorite_count': favorite,
                   'retweet_count': retweet})
        
twitter_plus = pd.DataFrame(df, columns = ['tweet_id', 'favorite_count', 'retweet_count'])

## Assessing Data


In [14]:
twitter

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
5     891087950875897856                    NaN                  NaN   
6     890971913173991426                    NaN                  NaN   
7     890729181411237888                    NaN                  NaN   
8     890609185150312448                    NaN                  NaN   
9     890240255349198849                    NaN                  NaN   
10    890006608113172480                    NaN                  NaN   
11    889880896479866881                    NaN                  NaN   
12    889665388333682689                    NaN                  NaN   
13    889638837579907072                    NaN                  NaN   
14    889531135344209921                    NaN                  NaN   
15    889278841981685760                    NaN                  NaN   
16    888917238123831296                    NaN                  NaN   
17    888804989199671297                    NaN                  NaN   
18    888554962724278272                    NaN                  NaN   
19    888202515573088257                    NaN                  NaN   
20    888078434458587136                    NaN                  NaN   
21    887705289381826560                    NaN                  NaN   
22    887517139158093824                    NaN                  NaN   
23    887473957103951883                    NaN                  NaN   
24    887343217045368832                    NaN                  NaN   
25    887101392804085760                    NaN                  NaN   
26    886983233522544640                    NaN                  NaN   
27    886736880519319552                    NaN                  NaN   
28    886680336477933568                    NaN                  NaN   
29    886366144734445568                    NaN                  NaN   
...                  ...                    ...                  ...   
2326  666411507551481857                    NaN                  NaN   
2327  666407126856765440                    NaN                  NaN   
2328  666396247373291520                    NaN                  NaN   
2329  666373753744588802                    NaN                  NaN   
2330  666362758909284353                    NaN                  NaN   
2331  666353288456101888                    NaN                  NaN   
2332  666345417576210432                    NaN                  NaN   
2333  666337882303524864                    NaN                  NaN   
2334  666293911632134144                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   
2336  666273097616637952                    NaN                  NaN   
2337  666268910803644416                    NaN                  NaN   
2338  666104133288665088                    NaN                  NaN   
2339  666102155909144576                    NaN                  NaN   
2340  666099513787052032                    NaN                  NaN   
2341  666094000022159362                    NaN                  NaN   
2342  666082916733198337                    NaN                  NaN   
2343  666073100786774016                    NaN                  NaN   
2344  666071193221509120                    NaN                  NaN   
2345  666063827256086533                    NaN                  NaN   
2346  666058600524156928                    NaN                  NaN   
2347  666057090499244032                    NaN                  NaN   
2348  666055525042405380                    NaN                  NaN   
2349  666051853826850816                    NaN         

In [15]:
image_prd

tweet_id                                            jpg_url  \
0     666020888022790149    https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1     666029285002620928    https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2     666033412701032449    https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3     666044226329800704    https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4     666049248165822465    https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
5     666050758794694657    https://pbs.twimg.com/media/CT5Jof1WUAEuVxN.jpg   
6     666051853826850816    https://pbs.twimg.com/media/CT5KoJ1WoAAJash.jpg   
7     666055525042405380    https://pbs.twimg.com/media/CT5N9tpXIAAifs1.jpg   
8     666057090499244032    https://pbs.twimg.com/media/CT5PY90WoAAQGLo.jpg   
9     666058600524156928    https://pbs.twimg.com/media/CT5Qw94XAAA_2dP.jpg   
10    666063827256086533    https://pbs.twimg.com/media/CT5Vg_wXIAAXfnj.jpg   
11    666071193221509120    https://pbs.twimg.com/media/CT5cN_3WEAAlOoZ.jpg   
12    666073100786774016    https://pbs.twimg.com/media/CT5d9DZXAAALcwe.jpg   
13    666082916733198337    https://pbs.twimg.com/media/CT5m4VGWEAAtKc8.jpg   
14    666094000022159362    https://pbs.twimg.com/media/CT5w9gUW4AAsBNN.jpg   
15    666099513787052032    https://pbs.twimg.com/media/CT51-JJUEAA6hV8.jpg   
16    666102155909144576    https://pbs.twimg.com/media/CT54YGiWUAEZnoK.jpg   
17    666104133288665088    https://pbs.twimg.com/media/CT56LSZWoAAlJj2.jpg   
18    666268910803644416    https://pbs.twimg.com/media/CT8QCd1WEAADXws.jpg   
19    666273097616637952    https://pbs.twimg.com/media/CT8T1mtUwAA3aqm.jpg   
20    666287406224695296    https://pbs.twimg.com/media/CT8g3BpUEAAuFjg.jpg   
21    666293911632134144    https://pbs.twimg.com/media/CT8mx7KW4AEQu8N.jpg   
22    666337882303524864    https://pbs.twimg.com/media/CT9OwFIWEAMuRje.jpg   
23    666345417576210432    https://pbs.twimg.com/media/CT9Vn7PWoAA_ZCM.jpg   
24    666353288456101888    https://pbs.twimg.com/media/CT9cx0tUEAAhNN_.jpg   
25    666362758909284353    https://pbs.twimg.com/media/CT9lXGsUcAAyUFt.jpg   
26    666373753744588802    https://pbs.twimg.com/media/CT9vZEYWUAAlZ05.jpg   
27    666396247373291520    https://pbs.twimg.com/media/CT-D2ZHWIAA3gK1.jpg   
28    666407126856765440    https://pbs.twimg.com/media/CT-NvwmW4AAugGZ.jpg   
29    666411507551481857    https://pbs.twimg.com/media/CT-RugiWIAELEaq.jpg   
...                  ...                                                ...   
2045  886366144734445568    https://pbs.twimg.com/media/DE0BTnQUwAApKEH.jpg   
2046  886680336477933568    https://pbs.twimg.com/media/DE4fEDzWAAAyHMM.jpg   
2047  886736880519319552    https://pbs.twimg.com/media/DE5Se8FXcAAJFx4.jpg   
2048  886983233522544640    https://pbs.twimg.com/media/DE8yicJW0AAAvBJ.jpg   
2049  887101392804085760    https://pbs.twimg.com/media/DE-eAq6UwAA-jaE.jpg   
2050  887343217045368832  https://pbs.twimg.com/ext_tw_video_thumb/88734...   
2051  887473957103951883    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2052  887517139158093824  https://pbs.twimg.com/ext_tw_video_thumb/88751...   
2053  887705289381826560    https://pbs.twimg.com/media/DFHDQBbXgAEqY7t.jpg   
2054  888078434458587136    https://pbs.twimg.com/media/DFMWn56WsAAkA7B.jpg   
2055  888202515573088257    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2056  888554962724278272    https://pbs.twimg.com/media/DFTH_O-UQAACu20.jpg   
2057  888804989199671297    https://pbs.twimg.com/media/DFWra-3VYAA2piG.jpg   
2058  888917238123831296    https://pbs.twimg.com/media/DFYRgsOUQAARGhO.jpg   
2059  889278841981685760  https://pbs.twimg.com/ext_tw_video_thumb/88927...   
2060  889531135344209921    https://pbs.twimg.com/media/DFg_2PVW0AEHN3p.jpg   
2061  889638837579907072    https://pbs.twimg.com/media/DFihzFfXsAYGDPR.jpg   
2062  889665388333682689    https://pbs.twimg.com/media/DFi579UWsAAatzw.jpg   
2063  889880896479866881    https://pbs.twimg.com/media/DFl99B1WsA

In [16]:
twitter_plus

tweet_id  favorite_count  retweet_count
0     892420643555336193           37414           8129
1     892177421306343426           32152           6025
2     891815181378084864           24248           3988
3     891689557279858688           40724           8294
4     891327558926688256           38952           8983
5     891087950875897856           19566           2984
6     890971913173991426           11432           1977
7     890729181411237888           63087          18092
8     890609185150312448           26921           4093
9     890240255349198849           30862           7071
10    890006608113172480           29673           7021
11    889880896479866881           26920           4778
12    889665388333682689           46493           9611
13    889638837579907072           26200           4340
14    889531135344209921           14625           2162
15    889278841981685760           24411           5155
16    888917238123831296           28151           4325
17    888804989199671297           24732           4104
18    888554962724278272           19168           3384
19    888078434458587136           21043           3337
20    887705289381826560           29197           5158
21    887517139158093824           44791          11221
22    887473957103951883           66613          17391
23    887343217045368832           32567          10000
24    887101392804085760           29575           5714
25    886983233522544640           33945           7422
26    886736880519319552           11627           3126
27    886680336477933568           21696           4280
28    886366144734445568           20471           3064
29    886267009285017600             116              4
...                  ...             ...            ...
2304  666411507551481857             426            319
2305  666407126856765440             103             37
2306  666396247373291520             161             81
2307  666373753744588802             179             87
2308  666362758909284353             753            552
2309  666353288456101888             211             71
2310  666345417576210432             282            131
2311  666337882303524864             190             88
2312  666293911632134144             484            341
2313  666287406224695296             141             63
2314  666273097616637952             167             75
2315  666268910803644416              99             34
2316  666104133288665088           13879           6281
2317  666102155909144576              75             11
2318  666099513787052032             148             64
2319  666094000022159362             158             69
2320  666082916733198337             110             43
2321  666073100786774016             311            155
2322  666071193221509120             140             55
2323  666063827256086533             458            209
2324  666058600524156928             108             56
2325  666057090499244032             282            135
2326  666055525042405380             422            230
2327  666051853826850816            1184            822
2328  666050758794694657             129             56
2329  666049248165822465             103             41
2330  666044226329800704             287            135
2331  666033412701032449             120             43
2332  666029285002620928             124             46
2333  666020888022790149            2491            489

[2334 rows x 3 columns]

In [17]:
twitter.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [18]:
image_prd.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [19]:
twitter_plus.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2334 entries, 0 to 2333
Data columns (total 3 columns):
tweet_id          2334 non-null int64
favorite_count    2334 non-null int64
retweet_count     2334 non-null int64
dtypes: int64(3)
memory usage: 54.8 KB


In [20]:
twitter.duplicated().sum(), image_prd.duplicated().sum(), twitter_plus.duplicated().sum()

(0, 0, 0)

In [48]:
image_prd['p1'].sample(20)

1635                       chow
661                     whippet
110                  Pomeranian
1960                    Samoyed
258          Labrador_retriever
2017         miniature_pinscher
1634               Irish_setter
208          miniature_pinscher
384                   Chihuahua
789                       Lhasa
28      black-and-tan_coonhound
1632             Siberian_husky
1541                   malamute
1016                        bib
979            Blenheim_spaniel
1967         Labrador_retriever
402                   porcupine
576                        maze
1015                  Chihuahua
1534         Labrador_retriever
Name: p1, dtype: object

In [49]:
twitter.shape[0], image_prd.shape[0], twitter_plus.shape[0]

(2356, 2075, 2334)

### Quality

**twitter**
- Missing some `expanded_urls`.
- Many names aren't real names, sometimes is just a letter or syllable.
- Data type is wrong `in_reply_to_status_id`, `in_reply_to_user_id`, `retweeted_status_id`, `retweeted_status_user_id` should be interger type and `timestamp`, `retweeted_status_timestamp` should be date type.
- Discard retweets because we are interested just in original ratings.
- Discard `source` column, because isn't necessary for this analysis.

**image_prd**
- Standardize `p1`,`p2`,`p3` columns to lowercase.
- Use underline as the standard separator.

**twitter_plus**
- Missing data, **twitter** has 2356 and **twitter_plus** has 2334.

### Tidyness

- Join the dataframe **twitter** with **twitter_plus**. 

**twitter**
- Convert the four columns `doggo`,`floofer`,`pupper` and `puppo` in just one called `stage`.